sdk/php/src/ValueObject/Argument.php

```diff
@@ -18,7 +18,24 @@ public function __construct(
         public string $description,
         public ListOfType|Type $type,
         public ?Json $default = null,
+        public ?string $defaultPath = null,
+        public ?array $ignore = null,
     ) {
+        if (!$type->nullable && $this->default == new Json('null')) {
+            throw new RuntimeException(sprintf(
+                'Non-nullable argument "%s" should not default to null.' .
+                ' This error should only occur if constructed manually.' .
+                ' Otherwise, it is a bug.',
+                $this->name,
+            ));
+        }
+    }
+
+    public function isOptional(): bool
+    {
+        return $this->type->nullable
+            || $this->default !== null
+            || $this->defaultPath !== null;
     }
```

Comment:
The `TypeDef.withOptional` API actually means "does it take null values?" See:
- https://github.com/dagger/dagger/issues/6749

It's best not to propagate that into the SDK, in this case by renaming the function more accurately. Then you'll see that having a default doesn't belong here:

```suggestion
    public function isNullable(): bool
    {
        return $this->type->nullable
            || $this->defaultPath !== null;
    }
```

But now you can also see that having a `defaultPath` doesn't really make it nullable. In fact this is an implementation detail leak from the engine and we should force it there instead of in the SDKs. So now if you remove that too, you're left with:

```php
    public function isNullable(): bool
    {
        return $this->type->nullable;
    }
```

And there's not much point in that 🙂

For now, you can just remove the `default !== null` part, and I'll move the `defaultPath` part to the engine soon.

Reply:
@helderco I see, thank you for the detailed feedback!

I think I have addressed this now, please let me know if it looks okay.

Since it is more of a temporary fix and the method would eventually be redundant; I've removed it, in favour of just checking it directly in the entrypoint, with a todo to remove it once handled by the engine:

https://github.com/dagger/dagger/blob/b9041a4d7eabfa2b5802b7f903589324f1b343be/sdk/php/src/Command/EntrypointCommand.php#L75-L82

Reply:
Fyi:
- https://github.com/dagger/dagger/pull/8572